In [293]:
import pandas as pd
import numpy as np
import sklearn
import json
import time
import os
from sklearn.ensemble import RandomForestRegressor
from sklearn.model_selection import train_test_split

In [2]:
# print("Printed immediately.")
# time.sleep(2.4)
# print("Printed after 2.4 seconds.")

In [3]:
# with open('data.txt', 'w') as outfile:
#     json.dump(x, outfile)

In [4]:
# with open('data.txt') as json_file:
#     data = json.load(json_file)

In [5]:
# for i in range(len(dates_from)):
#     data = search(dates_from[i], dates_to[i], 5)
#     with open(f'data/data{i}.txt', 'w') as outfile:
#         json.dump(data, outfile)
#     time.sleep(5)

In [6]:
# def job_id_extract(data):
#     list = []
#     for i in data['result']:
#         list.append(i['job_id'])
        
#     return(list)

In [ ]:
# my_dict = {}
# for i in range(len(dates_from)):
#     with open(f'data/data{i}.txt') as json_file:
#         my_dict[i] = json.load(json_file)

In [27]:
dates_from = ['2020-09-28 00:00', '2020-09-29 00:00', '2020-9-30 00:00', '2020-10-01 00:00', '2020-10-02 00:00', '2020-10-03 00:00', '2020-10-04 00:00', '2020-10-05 00:00', '2020-10-06 00:00', '2020-10-07 00:00', '2020-10-08 00:00', '2020-10-09 00:00', '2020-10-10 00:00', '2020-10-11 00:00', '2020-10-12 00:00', '2020-10-13 00:00', '2020-10-14 00:00', '2020-10-15 00:00', '2020-10-16 00:00', '2020-10-17 00:00', '2020-10-18 00:00', '2020-10-19 00:00', '2020-10-20 00:00']
dates_to = ['2020-09-29 00:00', '2020-09-30 00:00', '2020-10-01 00:00', '2020-10-02 00:00', '2020-10-03 00:00', '2020-10-04 00:00', '2020-10-05 00:00', '2020-10-06 00:00', '2020-10-07 00:00', '2020-10-08 00:00', '2020-10-09 00:00', '2020-10-10 00:00', '2020-10-11 00:00', '2020-10-12 00:00', '2020-10-13 00:00', '2020-10-14 00:00', '2020-10-15 00:00', '2020-10-16 00:00', '2020-10-17 00:00', '2020-10-18 00:00', '2020-10-19 00:00', '2020-10-20 00:00', '2020-10-21 00:00']

# Introduction, Business Understanding

Questions:
- How does Hybrid Analysis construct its threat score for URL's?
    - Can a generalized model be made based on the reports their Falcon Sandbox platform produces?
- What features are important to Hybrid Analysis URL threat score?
    - Can these features be determined, and then in turn, easily be extracted with guidance from their reports.
    - If reverse engineering isn't possible, can generealized insights still be produced?
    
Objectives:
- Creae a model that interprets Hybrid Analysis reports to determine threat score on URL's
- Create a feature list of Hybrid Analysis most informative features
- If the above are successful, deploy a reverse engineered model

Success Criteria:
- Either a functional model or insights into Hybrid Analysis feature importance for URL's that could guide another project

Hybrid Analysis:
- a website that allows users to submit malware, url's executables, html, etc for a proprietary "threat score" analysis.

# What have I done so far?

made search query and writing to file functions
made job_id extracting function that takes a dictionary input

can I get the things I need?
I need
- large set of malicious URL's
    - I can get these using my search and write function using a large enough window
- the malicious URLS's accompanying JOB id's
    - I can get these by extraccting the job ID's using my function on the dictionary
- Now I have job id's, target values, now I need features.
    - I need to use report get to get reports for each of my malicios URLS's I am storing that have the features
- Researching what is exactly contained in the feature reports
    - MUTEX (Mutual Exclusion) Locking down parts of code by keeping it in use, sometimes bloking visibility of other programs and making it immutable to outside other programs, think "file is in use somewhere else, cannot delete"

In [263]:
from search_query import search
from report_get import report
from job_id_extract import job_id_extract
from search_write_to_file import search_and_write
from dict_make import dict_make
from report_get_write_to_file import report_write
from target_extract import threat_score_extract

In [2]:
# dates_from = ['2020-09-28 00:00', '2020-09-29 00:00', '2020-9-30 00:00', '2020-10-01 00:00', '2020-10-02 00:00', '2020-10-03 00:00', '2020-10-04 00:00', '2020-10-05 00:00', '2020-10-06 00:00', '2020-10-07 00:00', '2020-10-08 00:00', '2020-10-09 00:00', '2020-10-10 00:00', '2020-10-11 00:00', '2020-10-12 00:00', '2020-10-13 00:00', '2020-10-14 00:00', '2020-10-15 00:00', '2020-10-16 00:00', '2020-10-17 00:00', '2020-10-18 00:00', '2020-10-19 00:00', '2020-10-20 00:00','2020-10-21 00:00', '2020-10-22 00:00', '2020-10-23 00:00', '2020-10-24 00:00', '2020-10-25 00:00', '2020-10-26 00:00', '2020-10-27 00:00', '2020-10-28 00:00','2020-10-29 00:00', '2020-11-01 00:00', '2020-11-02 00:00', '2020-11-03 00:00', '2020-11-04 00:00', '2020-11-05 00:00', '2020-11-06 00:00', '2020-11-07 00:00', '2020-11-08 00:00', '2020-11-09 00:00', '2020-11-10 00:00', '2020-11-11 00:00', '2020-11-12 00:00', '2020-11-13 00:00', '2020-11-14 00:00', '2020-11-15 00:00', '2020-11-16 00:00', '2020-11-17 00:00', '2020-11-18 00:00', '2020-11-19 00:00', '2020-11-20 00:00', '2020-11-21 00:00', '2020-11-22 00:00', '2020-11-23 00:00', '2020-11-24 00:00', '2020-11-25 00:00']
# dates_to = ['2020-09-29 00:00', '2020-09-30 00:00', '2020-10-01 00:00', '2020-10-02 00:00', '2020-10-03 00:00', '2020-10-04 00:00', '2020-10-05 00:00', '2020-10-06 00:00', '2020-10-07 00:00', '2020-10-08 00:00', '2020-10-09 00:00', '2020-10-10 00:00', '2020-10-11 00:00', '2020-10-12 00:00', '2020-10-13 00:00', '2020-10-14 00:00', '2020-10-15 00:00', '2020-10-16 00:00', '2020-10-17 00:00', '2020-10-18 00:00', '2020-10-19 00:00', '2020-10-20 00:00', '2020-10-21 00:00', '2020-10-22 00:00', '2020-10-23 00:00', '2020-10-24 00:00', '2020-10-25 00:00', '2020-10-26 00:00', '2020-10-27 00:00', '2020-10-28 00:00','2020-10-29 00:00', '2020-11-01 00:00', '2020-11-02 00:00', '2020-11-03 00:00', '2020-11-04 00:00', '2020-11-05 00:00', '2020-11-06 00:00', '2020-11-07 00:00', '2020-11-08 00:00', '2020-11-09 00:00', '2020-11-10 00:00', '2020-11-11 00:00', '2020-11-12 00:00', '2020-11-13 00:00', '2020-11-14 00:00', '2020-11-15 00:00', '2020-11-16 00:00', '2020-11-17 00:00', '2020-11-18 00:00', '2020-11-19 00:00', '2020-11-20 00:00', '2020-11-21 00:00', '2020-11-22 00:00', '2020-11-23 00:00', '2020-11-24 00:00', '2020-11-25 00:00', '2020-11-26 00:00']

In [11]:
len(search('2020-09-28 00:00', '2020-09-29 00:00', 5))

3

In [9]:
# search_and_write(dates_from, dates_to)

In [253]:
with open('data/data0.txt') as json_file:
    data = json.load(json_file)
data

{'search_terms': [{'id': 'filetype', 'value': 'url'},
  {'id': 'env_id', 'value': '100'},
  {'id': 'verdict', 'value': '5'},
  {'id': 'date_from', 'value': '2020-09-28 00:00'},
  {'id': 'date_to', 'value': '2020-09-29 00:00'}],
 'count': 4,
 'result': [{'verdict': 'malicious',
   'av_detect': '2',
   'threat_score': 100,
   'vx_family': 'Malicious site',
   'job_id': '5f72774a50643e7bc7460603',
   'sha256': '18b965df088347aceff0a2e6cfd5abfe9150df9b529d902b0a28125efec25b33',
   'environment_id': 100,
   'analysis_start_time': '2020-09-28 23:52:45',
   'submit_name': 'httpchuyengiadienmay.webdungsan.com.url',
   'environment_description': 'Windows 7 32 bit',
   'size': 63,
   'type': None,
   'type_short': 'url'},
  {'verdict': 'malicious',
   'av_detect': '0',
   'threat_score': 90,
   'vx_family': None,
   'job_id': '5f7274e3e6faf447dd379eca',
   'sha256': 'abd1ab47f8f3e0da5060d9e54c46e2959d9d53d4e13903d97d8c2f1198968ce3',
   'environment_id': 100,
   'analysis_start_time': '2020-09-28

In [16]:
range(len(dates_from))

range(0, 54)

In [90]:
my_dict = dict_make(dates_from)

In [24]:
len(my_dict)

23

In [25]:
my_dict

{0: {'search_terms': [{'id': 'filetype', 'value': 'url'},
   {'id': 'env_id', 'value': '100'},
   {'id': 'verdict', 'value': '5'},
   {'id': 'date_from', 'value': '2020-09-28 00:00'},
   {'id': 'date_to', 'value': '2020-09-29 00:00'}],
  'count': 4,
  'result': [{'verdict': 'malicious',
    'av_detect': '2',
    'threat_score': 100,
    'vx_family': 'Malicious site',
    'job_id': '5f72774a50643e7bc7460603',
    'sha256': '18b965df088347aceff0a2e6cfd5abfe9150df9b529d902b0a28125efec25b33',
    'environment_id': 100,
    'analysis_start_time': '2020-09-28 23:52:45',
    'submit_name': 'httpchuyengiadienmay.webdungsan.com.url',
    'environment_description': 'Windows 7 32 bit',
    'size': 63,
    'type': None,
    'type_short': 'url'},
   {'verdict': 'malicious',
    'av_detect': '0',
    'threat_score': 90,
    'vx_family': None,
    'job_id': '5f7274e3e6faf447dd379eca',
    'sha256': 'abd1ab47f8f3e0da5060d9e54c46e2959d9d53d4e13903d97d8c2f1198968ce3',
    'environment_id': 100,
    'ana

In [66]:
len(range(len(dates_from)))

23

In [252]:
my_list = job_id_extract(my_dict)

In [24]:
with open('reports/reports5f7e47fdb8e47e68de0a152a.txt') as json_file:
    data = json.load(json_file)
data

{'response': {'uuid': 'java:java.util.UUID',
  'xmlns': 'http://www.misp-project.org/',
  'Event': {'id': 'aa4e6f40-f451-4e2e-8d62-e3bb89af86b0',
   'date': '2020-10-07',
   'info': 'Falcon Sandbox auto-generated for "f399f059fe85ad2f58fcfc24f4fd830d8877a768803dfa5dc25ca93b4c673c22.url"',
   'analysis': 2,
   'distribution': 1,
   'published': 1,
   'Attribute': [{'category': 'External analysis',
     'type': 'link',
     'value': 'https://www.hybrid-analysis.com/search?query=f399f059fe85ad2f58fcfc24f4fd830d8877a768803dfa5dc25ca93b4c673c22',
     'distribution': 1},
    {'category': 'External analysis',
     'type': 'comment',
     'value': 'Falcon Sandbox v8.43 Copyright 2019 Hybrid Analysis GmbH, All Rights Reserved, www.payload-security.com',
     'distribution': 1},
    {'category': 'Payload delivery',
     'type': 'filename|md5',
     'value': 'f399f059fe85ad2f58fcfc24f4fd830d8877a768803dfa5dc25ca93b4c673c22.url|40ce0241a6c915bf62bedd89aa73f4c0',
     'distribution': 1},
    {'cat

In [264]:
dict_of_threatscores = threat_score_extract(a_dict)

In [271]:
target = pd.Series(dict_of_threatscore)

In [272]:
target

5f72774a50643e7bc7460603    100
5f7274e3e6faf447dd379eca     90
5f72734166e01f097b2eeb44     55
5f7271bd57dba718d1561324     41
5f751bc17735bd73cb45719f     63
                           ... 
5f8f786fab21e820cb349369     63
5f8f786f2b6ec67ce2322643     63
5f8f786f1e6bbf117f3f48f9     63
5f8f7137dcd73c04135d12df     81
5f8f70c91e207069a76e4876     75
Length: 159, dtype: int64

In [23]:
len(my_list)

159

In [16]:
report_write(my_list)

In [17]:
rep2 = report(my_list[1])

In [18]:
rep2

{'response': {'uuid': 'java:java.util.UUID',
  'xmlns': 'http://www.misp-project.org/',
  'Event': {'id': 'ad1c1758-6eee-41e5-bf32-9469bb03964d',
   'date': '2020-09-28',
   'info': 'Falcon Sandbox auto-generated for "abd1ab47f8f3e0da5060d9e54c46e2959d9d53d4e13903d97d8c2f1198968ce3.url"',
   'analysis': 2,
   'distribution': 1,
   'published': 1,
   'Attribute': [{'category': 'External analysis',
     'type': 'link',
     'value': 'https://www.hybrid-analysis.com/search?query=abd1ab47f8f3e0da5060d9e54c46e2959d9d53d4e13903d97d8c2f1198968ce3',
     'distribution': 1},
    {'category': 'External analysis',
     'type': 'comment',
     'value': 'Falcon Sandbox v8.43 Copyright 2019 Hybrid Analysis GmbH, All Rights Reserved, www.payload-security.com',
     'distribution': 1},
    {'category': 'Payload delivery',
     'type': 'filename|md5',
     'value': 'abd1ab47f8f3e0da5060d9e54c46e2959d9d53d4e13903d97d8c2f1198968ce3.url|3ecdc8b5d1561cdc806e8c6f71a7628a',
     'distribution': 1},
    {'cat

In [134]:
rep = report("5f9a05839fb71b523c1a2b59")

In [142]:
rep

{'response': {'uuid': 'java:java.util.UUID',
  'xmlns': 'http://www.misp-project.org/',
  'Event': {'id': 'f36070de-bc52-45ce-84d4-d6db3c47de26',
   'date': '2020-10-29',
   'info': 'Falcon Sandbox auto-generated for "290605d4e1a27ba1bf5539f0390d4b7437fe0eac0260984cf4c5fa259fc97b11.url"',
   'analysis': 2,
   'distribution': 1,
   'published': 1,
   'Attribute': [{'category': 'External analysis',
     'type': 'link',
     'value': 'https://www.hybrid-analysis.com/search?query=290605d4e1a27ba1bf5539f0390d4b7437fe0eac0260984cf4c5fa259fc97b11',
     'distribution': 1},
    {'category': 'External analysis',
     'type': 'comment',
     'value': 'Falcon Sandbox v8.43 Copyright 2019 Hybrid Analysis GmbH, All Rights Reserved, www.payload-security.com',
     'distribution': 1},
    {'category': 'Payload delivery',
     'type': 'filename|md5',
     'value': '290605d4e1a27ba1bf5539f0390d4b7437fe0eac0260984cf4c5fa259fc97b11.url|c01e09696b3597b4d57b95fe4a54abc7',
     'distribution': 1},
    {'cat

In [30]:
len(rep['response']['Event']['Attribute'])

535

In [237]:
#count unique categories
category_list = []
for i in rep['response']['Event']['Attribute']:
    category_list.append(i['category'])
set(category_list)
# one_hot = []
# if 'Artifacts dropped' in set(list):
#     one_hot.append(1)
    
# else:
#     one_hot.append(0)
# one_hot

# for i in set(list):
#     print('this could be it')

{'Artifacts dropped',
 'External analysis',
 'Network activity',
 'Payload delivery',
 'Payload installation',
 'Persistence mechanism'}

In [192]:
#count unique types
clist = []
for i in rep['response']['Event']['Attribute']:
    clist.append(i['type'])
clist
set(clist)

{'comment',
 'domain|ip',
 'filename|md5',
 'filename|sha1',
 'filename|sha256',
 'filename|sha512',
 'ip-dst',
 'link',
 'mutex',
 'regkey|value',
 'user-agent'}

In [168]:
list_type = []
list_category = []
for i in my_list:
    with open(f'reports/reports{i}.txt') as json_file:
        features = []
        job_id = i
        x = json.load(json_file)
        
        for i in x['response']['Event']['Attribute']:
            list_category.append(i['category'])
        
        
        for i in x['response']['Event']['Attribute']:
            list_type.append(i['type'])
print(set(list_type), set(list_category))

{'filename|md5', 'user-agent', 'domain|ip', 'mutex', 'ip-dst', 'regkey|value', 'comment', 'filename|sha512', 'domain', 'filename|sha256', 'link', 'filename|sha1'} {'Payload installation', 'Artifacts dropped', 'Network activity', 'External analysis', 'Persistence mechanism', 'Payload delivery'}


In [195]:
p = set(clist) - set(list_type)
p

set()

In [88]:
# for i in os.listdir('reports'):
#     try:
#         with open(f'{i}') as json_file:
#             x = json.load(json_file)
#             print(x)
#     except:
#         print(i)

.ipynb_checkpoints
reports5b2c3c377ca3e11cf2403363.txt
reports5c63af4f7ca3e1442e0e5365.txt
reports5ce69e940388384b8bfcf4d3.txt
reports5d02b1970288383f43c592ba.txt
reports5d39da17028838c6133fff34.txt
reports5f368e7d15235f3c982b9a14.txt
reports5f3c35f1f56cbf028005873d.txt
reports5f63f6ae1b224b45e91fd75f.txt
reports5f7271bd57dba718d1561324.txt
reports5f72734166e01f097b2eeb44.txt
reports5f7274e3e6faf447dd379eca.txt
reports5f72774a50643e7bc7460603.txt
reports5f7518e77ce8d170f2026b13.txt
reports5f7519b4eb07ca5d2737a6ba.txt
reports5f751bc17735bd73cb45719f.txt
reports5f76636e8fcd640116027e67.txt
reports5f7666b45acd7821452da3e2.txt
reports5f766718ddd5e2412c50f1b5.txt
reports5f7667707c1d470c4b087d18.txt
reports5f76685a7266f93be71c56d5.txt
reports5f766bbead0d270b480e418a.txt
reports5f77a46dbe523545dc4eef6c.txt
reports5f77bb5ebfbd29409973936e.txt
reports5f790e789db62354fa4c8326.txt
reports5f7a607696ec4c562f074bfb.txt
reports5f7d0472a7ac11745a048fa8.txt
reports5f7d5bb274d3c55a4703f81b.txt
reports5f

In [224]:
# can open all files in reports - this will become my core feature extraction
a_dict = {}
category_list = ['Artifacts dropped', 'External analysis', 'Network activity', 'Payload delivery', 'Payload installation', 'Persistence mechanism']
types_list = ['filename|md5', 'user-agent', 'domain|ip', 'mutex', 'ip-dst', 'regkey|value', 'comment', 'filename|sha512', 'domain', 'filename|sha256', 'link', 'filename|sha1']
for i in my_list:
    with open(f'reports/reports{i}.txt') as json_file:
        features = []
        job_id = i
        x = json.load(json_file)
        
        
        #acg length of attribute vlaues
        lenRep = len(x['response']['Event']['Attribute'])
        list_of_lengths = []
        sum = 0
        
        for i in x['response']['Event']['Attribute']:
            list_of_lengths.append(i['value'])
        for i in list_of_lengths:
            sum += len(i)
        avg_len = (sum/lenRep)
        
        features.append(avg_len)
        
        
        # total length of attribute values
        list_of_lengths = []
        sum = 0
        
        for i in x['response']['Event']['Attribute']:
            list_of_lengths.append(i['value'])
        for i in list_of_lengths:
            sum += len(i)
        
        features.append(sum)
        
        # count of categories
        list = []
        for i in x['response']['Event']['Attribute']:
            list.append(i['category'])
        len_list = len(list)
        
        features.append(len_list)
        
        # one hotting categories
        cats = []
        for i in x['response']['Event']['Attribute']:
            cats.append(i['category'])
        cats = set(cats)
        for i in category_list:
            if i in cats:
                features.append(1)
            else:
                features.append(0)
        
        types = []
        for i in x['response']['Event']['Attribute']:
            types.append(i['type'])
        types = set(types)
        for i in types_list:
            if i in types:
                features.append(1)
            else:
                features.append(0)
        
        
        a_dict[f'{job_id}'] = features
a_dict

{'5f72774a50643e7bc7460603': [102.38214285714285,
  86001,
  840,
  1,
  1,
  1,
  1,
  1,
  1,
  1,
  1,
  1,
  1,
  1,
  1,
  1,
  1,
  0,
  1,
  1,
  1],
 '5f7274e3e6faf447dd379eca': [128.87673231218088,
  176690,
  1371,
  1,
  1,
  1,
  1,
  1,
  1,
  1,
  1,
  1,
  1,
  1,
  1,
  1,
  1,
  0,
  1,
  1,
  1],
 '5f72734166e01f097b2eeb44': [123.60211002260739,
  164020,
  1327,
  1,
  1,
  1,
  1,
  1,
  1,
  1,
  1,
  1,
  1,
  1,
  1,
  1,
  1,
  0,
  1,
  1,
  1],
 '5f7271bd57dba718d1561324': [117.14173228346456,
  59508,
  508,
  1,
  1,
  1,
  1,
  1,
  1,
  1,
  1,
  1,
  1,
  1,
  1,
  1,
  1,
  0,
  1,
  1,
  1],
 '5f751bc17735bd73cb45719f': [152.42264150943396,
  40392,
  265,
  1,
  1,
  1,
  1,
  1,
  1,
  1,
  0,
  1,
  1,
  1,
  1,
  1,
  1,
  1,
  1,
  1,
  1],
 '5f7519b4eb07ca5d2737a6ba': [149.07094594594594,
  44125,
  296,
  1,
  1,
  1,
  1,
  1,
  1,
  1,
  1,
  1,
  1,
  1,
  1,
  1,
  1,
  1,
  1,
  1,
  1],
 '5f7518e77ce8d170f2026b13': [125.68440366972477,
  68

In [284]:
df = pd.DataFrame(a_dict).T
cols = ['job_id','avg_val_len', 'total_val_len','category_count','Artifacts dropped', 'External analysis', 'Network activity', 'Payload delivery', 'Payload installation', 'Persistence mechanism', 'filename|md5', 'user-agent', 'domain|ip', 'mutex', 'ip-dst', 'regkey|value', 'comment', 'filename|sha512', 'domain', 'filename|sha256', 'link', 'filename|sha1']
df

0         1       2    3    4    5    6   \
5f72774a50643e7bc7460603  102.382143   86001.0   840.0  1.0  1.0  1.0  1.0   
5f7274e3e6faf447dd379eca  128.876732  176690.0  1371.0  1.0  1.0  1.0  1.0   
5f72734166e01f097b2eeb44  123.602110  164020.0  1327.0  1.0  1.0  1.0  1.0   
5f7271bd57dba718d1561324  117.141732   59508.0   508.0  1.0  1.0  1.0  1.0   
5f751bc17735bd73cb45719f  152.422642   40392.0   265.0  1.0  1.0  1.0  1.0   
...                              ...       ...     ...  ...  ...  ...  ...   
5f8f786fab21e820cb349369  150.880597   70763.0   469.0  1.0  1.0  1.0  1.0   
5f8f786f2b6ec67ce2322643  166.184154   77608.0   467.0  1.0  1.0  1.0  1.0   
5f8f786f1e6bbf117f3f48f9  153.259341   69733.0   455.0  1.0  1.0  1.0  1.0   
5f8f7137dcd73c04135d12df  121.845802   79809.0   655.0  1.0  1.0  1.0  1.0   
5f8f70c91e207069a76e4876  131.793956   95946.0   728.0  1.0  1.0  1.0  1.0   

                           7    8    9   ...   11   12   13   14   15   16  \
5f72774a50643e7bc7460603  1.0  1.0  1.0  ...  1.0  1.0  1.0  1.0  1.0  1.0   
5f7274e3e6faf447dd379eca  1.0  1.0  1.0  ...  1.0  1.0  1.0  1.0  1.0  1.0   
5f72734166e01f097b2eeb44  1.0  1.0  1.0  ...  1.0  1.0  1.0  1.0  1.0  1.0   
5f7271bd57dba718d1561324  1.0  1.0  1.0  ...  1.0  1.0  1.0  1.0  1.0  1.0   
5f751bc17735bd73cb45719f  1.0  1.0  1.0  ...  1.0  1.0  1.0  1.0  1.0  1.0   
...                       ...  ...  ...  ...  ...  ...  ...  ...  ...  ...   
5f8f786fab21e820cb349369  1.0  1.0  1.0  ...  1.0  1.0  1.0  1.0  1.0  1.0   
5f8f786f2b6ec67ce2322643  1.0  1.0  1.0  ...  1.0  1.0  1.0  1.0  1.0  1.0   
5f8f786f1e6bbf117f3f48f9  1.0  1.0  1.0  ...  1.0  1.0  1.0  1.0  1.0  1.0   
5f8f7137dcd73c04135d12df  1.0  1.0  1.0  ...  1.0  1.0  1.0  1.0  1.0  1.0   
5f8f70c91e207069a76e4876  1.0  1.0  1.0  ...  1.0  1.0  1.0  1.0  1.0  1.0   

                           17   18   19   20  
5f72774a50643e7bc7460603  0.0  1.0  1.0  1.0  
5f7274e3e6faf447dd379eca  0.0  1.0  1.0  1.0  
5f72734166e01f097b2eeb44  0.0  1.0  1.0  1.0  
5f7271bd57dba718d1561324  0.0  1.0  1.0  1.0  
5f751bc17735bd73cb45719f  1.0  1.0  1.0  1.0  
...                       ...  ...  ...  ...  
5f8f786fab21e820cb349369  0.0  1.0  1.0  1.0  
5f8f786f2b6ec67ce2322643  0.0  1.0  1.0  1.0  
5f8f786f1e6bbf117f3f48f9  0.0  1.0  1.0  1.0  
5f8f7137dcd73c04135d12df  0.0  1.0  1.0  1.0  
5f8f70c91e207069a76e4876  0.0  1.0  1.0  1.0  

[159 rows x 21 columns]

In [285]:
df = df.reset_index()
df.columns = cols

In [286]:
df

job_id  avg_val_len  total_val_len  category_count  \
0    5f72774a50643e7bc7460603   102.382143        86001.0           840.0   
1    5f7274e3e6faf447dd379eca   128.876732       176690.0          1371.0   
2    5f72734166e01f097b2eeb44   123.602110       164020.0          1327.0   
3    5f7271bd57dba718d1561324   117.141732        59508.0           508.0   
4    5f751bc17735bd73cb45719f   152.422642        40392.0           265.0   
..                        ...          ...            ...             ...   
154  5f8f786fab21e820cb349369   150.880597        70763.0           469.0   
155  5f8f786f2b6ec67ce2322643   166.184154        77608.0           467.0   
156  5f8f786f1e6bbf117f3f48f9   153.259341        69733.0           455.0   
157  5f8f7137dcd73c04135d12df   121.845802        79809.0           655.0   
158  5f8f70c91e207069a76e4876   131.793956        95946.0           728.0   

     Artifacts dropped  External analysis  Network activity  Payload delivery  \
0                  1.0                1.0               1.0               1.0   
1                  1.0                1.0               1.0               1.0   
2                  1.0                1.0               1.0               1.0   
3                  1.0                1.0               1.0               1.0   
4                  1.0                1.0               1.0               1.0   
..                 ...                ...               ...               ...   
154                1.0                1.0               1.0               1.0   
155                1.0                1.0               1.0               1.0   
156                1.0                1.0               1.0               1.0   
157                1.0                1.0               1.0               1.0   
158                1.0                1.0               1.0               1.0   

     Payload installation  Persistence mechanism  ...  domain|ip  mutex  \
0                     1.0                    1.0  ...        1.0    1.0   
1                     1.0                    1.0  ...        1.0    1.0   
2                     1.0                    1.0  ...        1.0    1.0   
3                     1.0                    1.0  ...        1.0    1.0   
4                     1.0                    1.0  ...        1.0    1.0   
..                    ...                    ...  ...        ...    ...   
154                   1.0                    1.0  ...        1.0    1.0   
155                   1.0                    1.0  ...        1.0    1.0   
156                   1.0                    1.0  ...        1.0    1.0   
157                   1.0                    1.0  ...        1.0    1.0   
158                   1.0                    1.0  ...        1.0    1.0   

     ip-dst  regkey|value  comment  filename|sha512  domain  filename|sha256  \
0       1.0           1.0      1.0              1.0     0.0              1.0   
1       1.0           1.0      1.0              1.0     0.0              1.0   
2       1.0           1.0      1.0              1.0     0.0              1.0   
3       1.0           1.0      1.0              1.0     0.0              1.0   
4       1.0           1.0      1.0              1.0     1.0              1.0   
..      ...           ...      ...              ...     ...              ...   
154     1.0           1.0      1.0              1.0     0.0              1.0   
155     1.0           1.0      1.0              1.0     0.0              1.0   
156     1.0           1.0      1.0              1.0     0.0              1.0   
157     1.0           1.0      1.0              1.0     0.0              1.0   
158     1.0           1.0      1.0              1.0     0.0              1.0   

     link  filename|sha1  
0     1.0            1.0  
1     1.0            1.0  
2     1.0            1.0  
3     1.0            1.0  
4     1.0            1.0  
..    ...            ...  
154   1.0            1.0  
155   1.0            1.0  
156   1.0           

In [275]:
target.name = 'target'

In [277]:
target

5f72774a50643e7bc7460603    100
5f7274e3e6faf447dd379eca     90
5f72734166e01f097b2eeb44     55
5f7271bd57dba718d1561324     41
5f751bc17735bd73cb45719f     63
                           ... 
5f8f786fab21e820cb349369     63
5f8f786f2b6ec67ce2322643     63
5f8f786f1e6bbf117f3f48f9     63
5f8f7137dcd73c04135d12df     81
5f8f70c91e207069a76e4876     75
Name: target, Length: 159, dtype: int64

In [287]:
df = df.join(target, 'job_id')

In [59]:
#avg length of categories values
lenRep = len(rep['response']['Event']['Attribute'])
list = []
sum = 0
for i in rep['response']['Event']['Attribute']:
    list.append(i['value'])
for i in list:
    sum += len(i)
sum/lenRep

143.13271028037383

[{'category': 'External analysis',
  'type': 'link',
  'value': 'https://www.hybrid-analysis.com/search?query=290605d4e1a27ba1bf5539f0390d4b7437fe0eac0260984cf4c5fa259fc97b11',
  'distribution': 1},
 {'category': 'External analysis',
  'type': 'comment',
  'value': 'Falcon Sandbox v8.43 Copyright 2019 Hybrid Analysis GmbH, All Rights Reserved, www.payload-security.com',
  'distribution': 1},
 {'category': 'Payload delivery',
  'type': 'filename|md5',
  'value': '290605d4e1a27ba1bf5539f0390d4b7437fe0eac0260984cf4c5fa259fc97b11.url|c01e09696b3597b4d57b95fe4a54abc7',
  'distribution': 1},
 {'category': 'Payload delivery',
  'type': 'filename|sha1',
  'value': '290605d4e1a27ba1bf5539f0390d4b7437fe0eac0260984cf4c5fa259fc97b11.url|4ab3c69ffbd7c44491bc9c5b688bba0b6984f353',
  'distribution': 1},
 {'category': 'Payload delivery',
  'type': 'filename|sha256',
  'value': '290605d4e1a27ba1bf5539f0390d4b7437fe0eac0260984cf4c5fa259fc97b11.url|290605d4e1a27ba1bf5539f0390d4b7437fe0eac0260984cf4c5fa25

In [64]:
rep['response']['Event']['Attribute'][2]

{'category': 'Payload delivery',
 'type': 'filename|md5',
 'value': '290605d4e1a27ba1bf5539f0390d4b7437fe0eac0260984cf4c5fa259fc97b11.url|c01e09696b3597b4d57b95fe4a54abc7',
 'distribution': 1}

In [ ]:
159

# features

need to save a job ID for any kind of feature extraction, probably create a dictionary? creating values for columns? 'count_unique_categories':5, 'total_categories':153

couunt unique categories: each report has categories of nefarious things a url possesses. reachable by rep['response']['Event']['Attribute'] -> i['category']

```
list = []
for i in rep['response']['Event']['Attribute']:
    list.append(i['category'])
list
set(list)
```

avg length of categories values: these are usually just a hash, but it's still information, found in rep['response']['Event']['Attribute']['value']
```
lenRep = len(rep['response']['Event']['Attribute'])
list = []
sum = 0
for i in rep['response']['Event']['Attribute']:
    list.append(i['value'])
for i in list:
    sum += len(i)
sum/lenRep
```

total length of category values : same as above but just sum
```
lenRep = len(rep['response']['Event']['Attribute'])
list = []
sum = 0
for i in rep['response']['Event']['Attribute']:
    list.append(i['value'])
for i in list:
    sum += len(i)
sum/lenRep
```

#### MUTEX IS A TYPE!!!

rep['response']['Event']['Attribute'][i] - the i here is arbitrary, but what this does is drills down to the features, the 'attributes' are the features, mutex, payload delivery, persistence mechanism etc. can just count the number of attributes. If you add ['value'] to the end of the statement it gets the values of i which is typically a something with a hash?

avg length of value in payload_delivery

- can just use both on any number of feature possibilities

max length of something/anything

distance between feature_id 1 and 2 != 2 and 3

# Functions

dict_make - makes a combined dictionary of all downloaded data files from using search_and_write

job_id_extract - loops through the dictionary created by dict_make to append the job_id to a list

report_write - takes the job_id list extracted from job_id_extract and queries hybrid_analysis for the report that contains features
report_get - the GET request used in report_write


search_write_to_file - passes the date lists to search_query and gets the returns that contain the threat scores, then saves them as dataX
search_query - the post request used in search_write_to_file

In [288]:
df = df.drop(columns = 'job_id')

In [289]:
df

avg_val_len  total_val_len  category_count  Artifacts dropped  \
0     102.382143        86001.0           840.0                1.0   
1     128.876732       176690.0          1371.0                1.0   
2     123.602110       164020.0          1327.0                1.0   
3     117.141732        59508.0           508.0                1.0   
4     152.422642        40392.0           265.0                1.0   
..           ...            ...             ...                ...   
154   150.880597        70763.0           469.0                1.0   
155   166.184154        77608.0           467.0                1.0   
156   153.259341        69733.0           455.0                1.0   
157   121.845802        79809.0           655.0                1.0   
158   131.793956        95946.0           728.0                1.0   

     External analysis  Network activity  Payload delivery  \
0                  1.0               1.0               1.0   
1                  1.0               1.0               1.0   
2                  1.0               1.0               1.0   
3                  1.0               1.0               1.0   
4                  1.0               1.0               1.0   
..                 ...               ...               ...   
154                1.0               1.0               1.0   
155                1.0               1.0               1.0   
156                1.0               1.0               1.0   
157                1.0               1.0               1.0   
158                1.0               1.0               1.0   

     Payload installation  Persistence mechanism  filename|md5  ...  mutex  \
0                     1.0                    1.0           1.0  ...    1.0   
1                     1.0                    1.0           1.0  ...    1.0   
2                     1.0                    1.0           1.0  ...    1.0   
3                     1.0                    1.0           1.0  ...    1.0   
4                     1.0                    1.0           1.0  ...    1.0   
..                    ...                    ...           ...  ...    ...   
154                   1.0                    1.0           1.0  ...    1.0   
155                   1.0                    1.0           1.0  ...    1.0   
156                   1.0                    1.0           1.0  ...    1.0   
157                   1.0                    1.0           1.0  ...    1.0   
158                   1.0                    1.0           1.0  ...    1.0   

     ip-dst  regkey|value  comment  filename|sha512  domain  filename|sha256  \
0       1.0           1.0      1.0              1.0     0.0              1.0   
1       1.0           1.0      1.0              1.0     0.0              1.0   
2       1.0           1.0      1.0              1.0     0.0              1.0   
3       1.0           1.0      1.0              1.0     0.0              1.0   
4       1.0           1.0      1.0              1.0     1.0              1.0   
..      ...           ...      ...              ...     ...              ...   
154     1.0           1.0      1.0              1.0     0.0              1.0   
155     1.0           1.0      1.0              1.0     0.0              1.0   
156     1.0           1.0      1.0              1.0     0.0              1.0   
157     1.0           1.0      1.0              1.0     0.0              1.0   
158     1.0           1.0      1.0              1.0     0.0              1.0   

     link  filename|sha1  target  
0     1.0            1.0     100  
1     1.0            1.0      90  
2     1.0            1.0      55  
3     1.0            1.0      41  
4     1.0            1.0      63  
..    ...            ...     ...  
154   1.0            1.0      63  
155   1.0            1.0      63  
156   1.0            1.0      63  
157   1.0            1.0      81  
158   1.0            1.0      75  

[159 rows x 22 columns]

In [294]:
X = df.drop(columns = 'target')
y = df['target']

In [302]:
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.40, random_state=5)

In [303]:
rfr = RandomForestRegressor(max_depth=6, random_state=5)

In [304]:
rfr.fit(X_train, y_train)

RandomForestRegressor(max_depth=6, random_state=5)

In [305]:
pred = rfr.predict(X_test)

In [306]:
error = y_test - pred

In [307]:
error

26     5.833435
87    -9.465071
71     0.048401
66     6.904859
93    -6.380505
         ...   
101    3.737162
122    0.048401
124    0.903614
155    0.372492
49     0.143005
Name: target, Length: 64, dtype: float64

In [309]:
sum = 0
for i in error:
    sum += i
sum

-122.57677627141302